firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java

```diff
@@ -298,19 +293,36 @@ public void onError(Exception error) {
     return new ConfigAutoFetch(httpURLConnection, configFetchHandler, listeners, retryCallback);
   }
 
-  // Kicks off Http stream listening and AutoFetching. The AutoFetch method listenForNotifications
-  // will be a blocking call until the Http stream closes, at which point we will try and
-  // reestablish the stream.
+  /**
+   * Open the realtime connection, begin listening for updates, and auto-fetch when an update is
+   * received.
+   *
+   * <p>If the connection is successful, this method will block on its thread while it reads the
+   * chunk-encoded HTTP body. When the connection closes, it attempts to reestablish the stream.
+   */
   @SuppressLint("VisibleForTests")
   synchronized void beginRealtimeHttpStream() {
-    if (canMakeHttpStreamConnection()) {
-      this.httpURLConnection = makeHttpConnection();
-      if (httpURLConnection != null) {
-        resetRetryParameters();
-        ConfigAutoFetch configAutoFetch = startAutoFetch(httpURLConnection);
-        configAutoFetch.listenForNotifications();
-        closeRealtimeHttpStream();
-      }
+    if (!canMakeHttpStreamConnection()) {
+      return;
+    }
+
+    try {
+      // Create the open the connection.
+      httpURLConnection = createRealtimeConnection();
+      httpURLConnection.connect();
+
+      // Reset the retries remaining if we opened the connection without an exception.
+      resetRetryParameters();
+
+      // Start listening for realtime notifications.
+      ConfigAutoFetch configAutoFetch = startAutoFetch(httpURLConnection);
+      configAutoFetch.listenForNotifications();
+    } catch (IOException e) {
+      propagateErrors(
+          new FirebaseRemoteConfigRealtimeUpdateStreamException(
+              "Exception connecting to realtime stream. Retrying..."));
+    } finally {
+      closeRealtimeHttpStream();
       retryHTTPConnection();
```

Comment:
i don't think we need to retry for the success case?

Reply:
Sorta the same as above - I believe this is done because the `listenForNotifications()` blocks while it reads the stream, so at this point we're re-opening the closed stream.

Reply:
ignore this comment then